File: src/main/java/com/arcao/feedback/FeedbackHelper.java
package com.arcao.feedback;

import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;

import com.arcao.feedback.collector.AccountInfoCollector;
import com.arcao.feedback.collector.AppInfoCollector;
import com.arcao.feedback.collector.BuildConfigCollector;
import com.arcao.feedback.collector.Collector;
import com.arcao.feedback.collector.ConfigurationCollector;
import com.arcao.feedback.collector.ConstantsCollector;
import com.arcao.feedback.collector.DisplayManagerCollector;
import com.arcao.feedback.collector.LogCatCollector;
import com.arcao.feedback.collector.MemoryCollector;
import com.arcao.feedback.collector.SharedPreferencesCollector;

import org.apache.commons.io.IOUtils;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.Collection;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

import timber.log.Timber;

public class FeedbackHelper {
	public static void sendFeedback(@NonNull Context context, @StringRes int resEmail, @StringRes int resSubject, @StringRes int resMessageText) {
		String subject = context.getString(resSubject, getApplicationName(context), getVersion(context));

		String email = context.getString(resEmail);

		Intent intent = new Intent(Intent.ACTION_SEND_MULTIPLE); // only e-mail apps
		intent.setType("plain/text");
		intent.putExtra(Intent.EXTRA_SUBJECT, subject);
		intent.putExtra(Intent.EXTRA_TEXT, context.getString(resMessageText));
		intent.putExtra(Intent.EXTRA_EMAIL, new String[] {email});
		intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

		try {
			createReport(context, FeedbackFileProvider.getReportFile(context));

			ArrayList<Uri> uris = new ArrayList<>();
			uris.add(FeedbackFileProvider.getReportFileUri());

			intent.putParcelableArrayListExtra(Intent.EXTRA_STREAM, uris);
		} catch (IOException e) {
			Timber.e(e, e.getMessage());
		}

		context.startActivity(Intent.createChooser(intent, null));
	}

	private static void createReport(Context context, File reportFile) throws IOException {
		ZipOutputStream zos = null;

		if (reportFile.exists()) {
			Timber.d("Report file " + reportFile + " already exist.");
			if (reportFile.delete()) {
				Timber.d("Report file removed.");
			}
		}

		Timber.d("Creating report to " + reportFile);
		try {
			zos = new ZipOutputStream(new FileOutputStream(reportFile));
			writeCollectors(zos, context);
		}
		finally {
			IOUtils.closeQuietly(zos);
		}

		Timber.d("Report created.");
	}

	private static void writeCollectors(ZipOutputStream zos, Context context) throws IOException {
		Collection<Collector> collectors = prepareCollectors(context);

		for (Collector collector : collectors) {
			zos.putNextEntry(new ZipEntry(collector.getName() + ".txt"));

			OutputStreamWriter writer = new OutputStreamWriter(zos, "UTF-8");
			writer.write(collector.toString());
			writer.flush();

			zos.closeEntry();
		}
	}

	private static Collection<Collector> prepareCollectors(Context context) {
		Collection<Collector> collectors = new ArrayList<>();

		collectors.add(new AppInfoCollector(context));
		collectors.add(new BuildConfigCollector());
		collectors.add(new ConfigurationCollector(context));
		collectors.add(new ConstantsCollector(Build.class, "BUILD"));
		collectors.add(new ConstantsCollector(Build.VERSION.class, "VERSION"));
		collectors.add(new MemoryCollector());
		collectors.add(new LogCatCollector(context));
		collectors.add(new SharedPreferencesCollector(context));
		collectors.add(new DisplayManagerCollector(context));
		collectors.add(new AccountInfoCollector(context));

		return collectors;
	}

	private static String getApplicationName(Context context) {
		int stringId = context.getApplicationInfo().labelRes;
		return context.getString(stringId);
	}

	private static String getVersion(Context context) {
		try {
			return context.getPackageManager().getPackageInfo(context.getPackageName(), 0).versionName;
		} catch (PackageManager.NameNotFoundException e) {
			Timber.e(e, e.getMessage());
			return "0.0";
		}
	}
}

File: app/src/main/java/com/arcao/feedback/FeedbackHelper.kt
package com.arcao.feedback

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Parcelable
import androidx.annotation.NonNull
import androidx.annotation.StringRes
import com.arcao.feedback.collector.Collector
import com.arcao.geocaching4locus.App
import com.arcao.geocaching4locus.base.coroutine.CoroutinesDispatcherProvider
import kotlinx.coroutines.withContext
import org.koin.standalone.KoinComponent
import org.koin.standalone.inject
import timber.log.Timber
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStreamWriter
import java.util.Stack
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream

// TODO convert to class
object FeedbackHelper : KoinComponent {
    private val app by inject<App>()
    private val collectors by inject<List<Collector>>(DEP_FEEDBACK_COLLECTORS)
    private val dispatcherProvider by inject<CoroutinesDispatcherProvider>()

    @JvmStatic
    suspend fun sendFeedback(@NonNull activity: Activity, @StringRes resEmail: Int, @StringRes resSubject: Int, @StringRes resMessageText: Int) =
        withContext(dispatcherProvider.computation) {
            val subject = activity.getString(resSubject, app.name, app.version)

            val email = activity.getString(resEmail)

            val intent = Intent(Intent.ACTION_SEND).apply {
                // only e-mail apps
                type = "plain/text"
                putExtra(Intent.EXTRA_SUBJECT, subject)
                putExtra(Intent.EXTRA_TEXT, activity.getString(resMessageText))
                putExtra(Intent.EXTRA_EMAIL, arrayOf(email))
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }

            try {
                createReport(FeedbackFileProvider.getReportFile(activity))

                val reportUri = FeedbackFileProvider.reportFileUri
                intent.putExtra(Intent.EXTRA_STREAM, reportUri)

                // grant read permission
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                } else {
                    val resInfoList = activity.packageManager
                        .queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY)

                    for (resolveInfo in resInfoList) {
                        val packageName = resolveInfo.activityInfo.packageName
                        activity.grantUriPermission(packageName, reportUri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
                    }
                }
            } catch (e: IOException) {
                Timber.e(e)
            }

            withContext(dispatcherProvider.main) {
                activity.startActivity(createEmailOnlyChooserIntent(activity, intent, null))
            }
        }

    @Throws(IOException::class)
    private suspend fun createReport(reportFile: File) {
        if (reportFile.exists()) {
            Timber.d("Report file $reportFile already exist.")
            if (reportFile.delete()) {
                Timber.d("Report file removed.")
            }
        }

        Timber.d("Creating report to %s", reportFile)
        ZipOutputStream(FileOutputStream(reportFile)).use { zos -> writeCollectors(zos) }
        Timber.d("Report created.")
    }

    @Throws(IOException::class)
    private suspend fun writeCollectors(zos: ZipOutputStream) {
        for (collector in collectors) {
            zos.putNextEntry(ZipEntry(collector.name + ".txt"))

            OutputStreamWriter(zos, "UTF-8").apply {
                write(
                    try {
                        collector()
                    } catch (ignored: Exception) {
                        // ignored
                        ""
                    }
                )
                flush()
            }

            zos.closeEntry()
        }
    }

    private fun createEmailOnlyChooserIntent(context: Context, source: Intent, chooserTitle: CharSequence?): Intent {
        val intents = Stack<Intent>()
        val emailIntent = Intent(Intent.ACTION_SENDTO, Uri.parse("mailto:info@domain.com"))

        context.packageManager.queryIntentActivities(emailIntent, 0).forEach { resolveInfo ->
            intents.add(Intent(source).setPackage(resolveInfo.activityInfo.packageName))
        }

        return if (!intents.isEmpty()) {
            Intent.createChooser(intents.removeAt(0), chooserTitle)
                .putExtra(Intent.EXTRA_INITIAL_INTENTS, intents.toTypedArray<Parcelable>())
        } else {
            Intent.createChooser(source, chooserTitle)
        }
    }
}
